[Justine Koh] iP

justinekoh opened the pull request:
# BoB
>  "Your mind is for having ideas, not holding them." - David Allen [source](https://dansilvestre.com/productivity-quotes)

BoB frees your mind of having to remember things you need to do. It's

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/justinekoh/ip/releases/tag/A-Jar).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Added greet, echo and exit functions. Changed chatbot name to Bob.

Commit:
Change scanner.next to scanner.nextLine

Commit:
Added list and add functionality.

Commit:
Added Task class and updated chatbot to utilize Tasks.
Added Mark and Unmark functionalities.

Commit:
Added subclasses of Task, including Deadline, Event and ToDo classes.
Edited Add functionality of chatbot to support the above 3 task types.

Commit:
Setup automatic testing of chatbot.

Commit:
Added some exception handling for incompatible user inputs.

Commit:
Added Delete functionality and associated error handling. Fixed some bugs when trying to add an event with invalid arguments.

zhengtaoJ commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
```
Consider adding header comments for the class and its public methods?

zhengtaoJ commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
+    protected String by;
```
Perhaps using a noun here will be more meaningful?

zhengtaoJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,106 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        String line = "    ______________________________\n";
```
Consider making this a class-level constant?

zhengtaoJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,106 @@
+        Scanner sc = new Scanner(System.in);
+        List<Task> list = new ArrayList<>();
+
+        String cmd = sc.nextLine().strip();
```
Consider using a full English word as a variable name here?

Commit:
Add persistent storage functionality and change name of program to Bob

Commit:
Change date and time from string to LocalDateTime type

Commit:
Merge branch 'branch-Level-8'

Commit:
Add Ui and TaskList classes to improve abstraction

Commit:
Add JUnit tests and additional abstractions

eugenechiaay commented on src/main/java/bob/Task.java:
```diff
@@ -0,0 +1,50 @@
+        return (this.isDone ? "X" : " ");
+    }
+
+    public void toggleStatus() {
+        this.isDone = !this.isDone;
+    }
```
hmm, seems like this method is not used anywhere, perhaps you'll want to remove it?

```suggestion
    
```

eugenechiaay replied:
hmm, seems like this method is not used anywhere, perhaps you'll want to remove it?

```suggestion
    
```

eugenechiaay commented on src/main/java/bob/Triplet.java:
```diff
@@ -0,0 +1,29 @@
+package bob;
+
+public class Triplet<T, U, V> {
```
Cool to see you using generics here and making a triplet class. However, it could be more useful to name this class more purposefully, "Task"?

eugenechiaay replied:
Cool to see you using generics here and making a triplet class. However, it could be more useful to name this class more purposefully, "Task"?

eugenechiaay commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,67 @@
-Hello from
- ____        _        
-|  _ \ _   _| | _____ 
-| | | | | | | |/ / _ \
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
```
Maybe you can add a similar visual here.

eugenechiaay reviewed (commented):
Overall, I found your code easy to read.

imrajsinghsandhu commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    /**
+     * Returns the size of the TaskList.
```
Nicely indented code, and following coding standards well. Like to see it! Keep it up!

imrajsinghsandhu commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,72 @@
+    public static void parse(String userInput, TaskList tasks, Storage storage) {
+        try {
+            String cmd = userInput.split(" ")[0].strip();
+            boolean tasksIsDiff = false;
```
perhaps renaming tasksIsDiff to isTaskDiff would be better as a boolean name?

imrajsinghsandhu commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,113 @@
+        }
+
+        return new TaskList(list);
+    }
+
+    /**
+     * Converts a given list of tasks into a String and writes it to the file associated with this Storage instance.
+     *
+     * @param tasks List of tasks to be written to file.
+     */
+    public void writeTaskListToFile(TaskList tasks) {
+        StringBuilder tasksString = new StringBuilder();
+        for (int i = 0; i < tasks.size(); i++) {
+            tasksString.append(tasks.get(i).generateSavedEntry());
+            tasksString.append("\n");
+        }
+        try {
+            FileWriter fileWriter = new FileWriter(file);
+            fileWriter.write(tasksString.toString());
+            fileWriter.close();
+        } catch (IOException e) {
+            System.out.println("Error writing to file!:( : " + e.getMessage());
+        }
+    }
+}
```
I like that you have well-written javadocs, it makes code easier to read and understand!

imrajsinghsandhu reviewed (commented):
Overall, I found your code easy to read for the most part, and see that you have written it well! LGTM, and keep up the good work.

Commit:
Add JavaDoc comments on all public classes and most public methods

Commit:
Improve coding standard to improve readability

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Add Find functionality

Commit:
Merge branch 'branch-Level-9'

Commit:
Correct Find functionality

Commit:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

Commit:
Configure Gradle for project

Commit:
Add checkstyle to project

Commit:
Add GUI to Bob

Commit:
Add assertions to check for necessary conditions in program flow

There are multiple conditions that must hold true at every point of the program flow. If any of these conditions do not hold true when they should, it will cause unexpected behaviour in the program.

Having assertions in the code to guarantee such conditions gives us confidence that the program is running smoothly with all necessary conditions met at all assertion checkpoints if no assertions fail.

Commit:
Add .yml file for Continuous Integration

Commit:
Improve code quality

Code quality is important for readability and to facilitate maintenance of code.

Long methods have been shortened, by abstracting out suitable functional blocks of code, and in accordance to SLAP. Nested indentations have been removed and cleaned up.

Updated most JavaDocs.

justinekoh committed:
Merge pull request #2 from justinekoh/branch-A-CI

Add .yml file for Continuous Integration

Commit:
Merge branch 'master' of https://github.com/justinekoh/ip

* 'master' of https://github.com/justinekoh/ip:
  Add .yml file for Continuous Integration

Commit:
Merge branch 'master' into branch-A-Assertions

* master:
  Add .yml file for Continuous Integration

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add .yml file for Continuous Integration

justinekoh committed:
Merge pull request #3 from justinekoh/branch-A-Assertions

branch-A-Assertions

Commit:
Merge branch 'master' of https://github.com/justinekoh/ip

* 'master' of https://github.com/justinekoh/ip:
  Add assertions to check for necessary conditions in program flow

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assertions to check for necessary conditions in program flow

justinekoh committed:
Merge pull request #4 from justinekoh/branch-A-CodeQuality

Improve code quality

Commit:
Merge branch 'master' of https://github.com/justinekoh/ip into branch-A-CodeQuality

* 'master' of https://github.com/justinekoh/ip:

Commit:
Merge branch 'master' of https://github.com/justinekoh/ip

* 'master' of https://github.com/justinekoh/ip:
  Improve code quality

Commit:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:

Commit:
Add command and archive functionalities

justinekoh committed:
Merge pull request #5 from justinekoh/branch-C-Archive

Add command and archive functionalities

Commit:
Improve GUI

Changed Bob and user photos, made Bob and user photos round, added title
to application

Commit:
Improve GUI

Increased size of window, disabled window resizing and fixed bugs with
displayed text.

Commit:
Change JavaFX version in DialogBox and MainWindow fxml files

Commit:
Add Ui.png and edit JavaDocs

justinekoh committed:
Set theme jekyll-theme-minimal

Commit:
Update README.md

justinekoh committed:
Update README.md

justinekoh committed:
Set theme jekyll-theme-cayman

justinekoh committed:
Set theme jekyll-theme-minimal

justinekoh committed:
Update README.md

justinekoh committed:
Update README.md

justinekoh committed:
Add sample run to README

justinekoh committed:
Correct typo

Commit:
Add startup UI photo

Commit:
Edit startup UI photo

justinekoh committed:
Update README.md

Commit:
Fix bugs and JavaDocs

Commit:
Add testcases for ToDo, Event, Deadline classes

Commit:
Update Ui.png and User Guide

Commit:
Fix bugs, update Ui.png and improve naming convention
